From b53ee495a2afd675110680a600450f0cd7142d43 Mon Sep 17 00:00:00 2001
From: Bogdan Purcareata <bogdan.purcareata@nxp.com>
Date: Thu, 28 Sep 2017 11:07:33 +0000
Subject: [PATCH 519/706] staging: fsl-dpaa2/mac: Free phydev on unbind

Stop polling and unregister / disconnect the phydev structure on DPAA2
mac driver unbind.

Signed-off-by: Bogdan Purcareata <bogdan.purcareata@nxp.com>
(cherry picked from commit 88846f742fcb4c4c0f0c3f915413c19b2f675be1)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/staging/fsl-dpaa2/mac/mac.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/staging/fsl-dpaa2/mac/mac.c b/drivers/staging/fsl-dpaa2/mac/mac.c
index b622630a3096..23a72f1d19bc 100644
--- a/drivers/staging/fsl-dpaa2/mac/mac.c
+++ b/drivers/staging/fsl-dpaa2/mac/mac.c
@@ -624,6 +624,15 @@ static int dpaa2_mac_remove(struct fsl_mc_device *mc_dev)
 {
 	struct device		*dev = &mc_dev->dev;
 	struct dpaa2_mac_priv	*priv = dev_get_drvdata(dev);
+	struct net_device	*netdev = priv->netdev;
+
+	dpaa2_mac_stop(netdev);
+
+	if (phy_is_pseudo_fixed_link(netdev->phydev))
+		fixed_phy_unregister(netdev->phydev);
+	else
+		phy_disconnect(netdev->phydev);
+	netdev->phydev = NULL;
 
 #ifdef CONFIG_FSL_DPAA2_MAC_NETDEVS
 	unregister_netdev(priv->netdev);
-- 
2.17.1

